servers/zms/src/main/java/com/yahoo/athenz/zms/DBService.java

```diff
@@ -6402,6 +6403,147 @@ void enforceGroupUserAuthorityRestrictions(final String domainName, final String
+        List<GroupMember> groupMembersWithUpdatedState;
+        GroupMember groupMember;
+        DomainGroupMember domainGroupMember;
+        Set<GroupMember> updatedGroups = new HashSet<>();
+
+        domainGroupMember = con.getPrincipalGroups(updatedUser.getFullName(), null);
+        if (!domainGroupMember.getMemberGroups().isEmpty()) {
+            for (GroupMember currentGroup : domainGroupMember.getMemberGroups()) {
+                groupMember = new GroupMember();
+                groupMember.setMemberName(updatedUser.getFullName());
+                if (suspended) {
+                    groupMember.setSystemDisabled(currentGroup.getSystemDisabled() & ~Principal.State.AUTHORITY_SYSTEM_SUSPENDED.getValue());
+                } else {
+                    groupMember.setSystemDisabled(currentGroup.getSystemDisabled() | Principal.State.AUTHORITY_SYSTEM_SUSPENDED.getValue());
+                }
+                groupMembersWithUpdatedState = new ArrayList<>();
+                groupMembersWithUpdatedState.add(groupMember);
+
+                // Following method does Audit entry as well
+                if (updateGroupMemberDisabledState(null, con, groupMembersWithUpdatedState, currentGroup.getDomainName(),
+                        currentGroup.getGroupName(), ZMSConsts.SYS_AUTH_MONITOR, AUDIT_REF, caller)) {
+                    updatedGroups.add(currentGroup);
+                }
+            }
+            for (GroupMember updatedGM : updatedGroups) {
+
+                // update our group and domain time-stamps, and invalidate local cache entry
+                con.updateGroupModTimestamp(updatedGM.getDomainName(), updatedGM.getGroupName());
+                con.updateDomainModTimestamp(updatedGM.getDomainName());
+                cacheStore.invalidate(updatedGM.getDomainName());
+            }
+        }
+    }
+
+    private void updateRoleMembershipsByPrincipalState(boolean suspended, String caller, ObjectStoreConnection con, Principal updatedUser) {
+        RoleMember roleMember;
+        List<RoleMember> roleMembersWithUpdatedState;
+        DomainRoleMember domainRoleMember;
+        Set<MemberRole> updatedRoles = new HashSet<>();
+
+        domainRoleMember = con.getPrincipalRoles(updatedUser.getFullName(), null);
+        if (!domainRoleMember.getMemberRoles().isEmpty()) {
+            for (MemberRole memberRole : domainRoleMember.getMemberRoles()) {
+                roleMember = new RoleMember();
+                roleMember.setMemberName(updatedUser.getFullName());
+                if (suspended) {
+                    roleMember.setSystemDisabled(memberRole.getSystemDisabled() & ~Principal.State.AUTHORITY_SYSTEM_SUSPENDED.getValue());
+                } else {
+                    roleMember.setSystemDisabled(memberRole.getSystemDisabled() | Principal.State.AUTHORITY_SYSTEM_SUSPENDED.getValue());
+                }
+                roleMembersWithUpdatedState = new ArrayList<>();
```

Comment:
no need to create an array list with one member all the time. Instead use Collections.singletonList()